feat: add support for loading rules from global and project-local .roo directories

samhvw8 opened the pull request:
<!--
Thank you for contributing to Roo Code!

Before submitting your PR, please ensure:
- It's linked to an approved GitHub Issue.
- You've reviewed our [Contributing Guidelines](../CONTRIBUTING.md).
-->

### Related GitHub Issue

<!-- Every PR MUST be linked to an approved issue. -->

Closes: #4978 <!-- Replace with the issue number, e.g., Closes: #123 -->

### Description

#### Summary
This PR introduces a new hierarchical configuration system that supports both global (`~/.roo/`) and project-local (`.roo/`) directories for organizing rules and configuration files. This enhancement provides better organization and allows users to define global defaults while maintaining project-specific overrides.

#### What's Changed

##### 🆕 New Features
- **Global configuration support**: Rules can now be stored in `~/.roo/rules/` directory that apply across all projects
- **Project-local overrides**: Project-specific rules in `.roo/rules/` take precedence over global rules
- **Mode-specific rules**: Both global and project-local mode-specific rules are supported via `rules-{mode}/` directories
- **Hierarchical loading**: Global rules are loaded first, then project-local rules are appended with clear labeling

##### 🏗️ Architecture Changes
- **New `roo-config` service**: Centralized service for managing .roo directory operations
  - `getGlobalRooDirectory()` - Platform-aware global .roo path resolution
  - `getRooDirectoriesForCwd()` - Ordered list of directories to check
  - `loadConfiguration()` - Unified configuration loading with merge support
  - Comprehensive file system utilities (`directoryExists`, `fileExists`, `readFileIfExists`)

- **Enhanced `custom-instructions` module**: Updated to use the new roo-config service
  - Supports both new .roo directory structure and legacy .roorules files
  - Clear labeling of global vs project-specific rules in output
  - Maintains backward compatibility with existing rule files

<!--
Briefly summarize the changes in this PR and how they address the linked issue.
The issue should cover the "what" and "why"; this section should focus on:
- The "how": key implementation details, design choices, or trade-offs made.
- Anything specific reviewers should pay attention to in this PR.
-->

### Test Procedure

<!--
Detail the steps to test your changes. This helps reviewers verify your work.
- How did you test this specific implementation? (e.g., unit tests, manual testing steps)
- How can reviewers reproduce your tests or verify the fix/feature?
- Include relevant testing environment details if applicable.
-->

### Pre-Submission Checklist

<!-- Go through this checklist before marking your PR as ready for review. -->

- [x] **Issue Linked**: This PR is linked to an approved GitHub Issue (see "Related GitHub Issue" above).
- [x] **Scope**: My changes are focused on the linked issue (one major feature/fix per PR).
- [x] **Self-Review**: I have performed a thorough self-review of my code.
- [x] **Testing**: New and/or updated tests have been added to cover my changes (if applicable).
- [x] **Documentation Impact**: I have considered if my changes require documentation updates (see "Documentation Updates" section below).
- [x] **Contribution Guidelines**: I have read and agree to the [Contributor Guidelines](/CONTRIBUTING.md).

### Screenshots / Videos

<!--
For UI changes, please provide before-and-after screenshots or a short video of the *actual results*.
This greatly helps in understanding the visual impact of your changes.
-->

### Documentation Updates

<!--
Does this PR necessitate updates to user-facing documentation?
- [ ] No documentation updates are required.
- [ ] Yes, documentation updates are required. (Please describe what needs to be updated or link to a PR in the docs repository).
-->

### Additional Notes

<!-- Add any other context, questions, or information for reviewers here. -->

### Get in Touch

<!--
Please provide your Discord username for reviewers or maintainers to reach you if they have questions about your PR
-->

<!-- ELLIPSIS_HIDDEN -->

----

> [!IMPORTANT]
> Adds support for hierarchical rule loading from global and project-local `.roo` directories, with tests and backward compatibility.
> 
>   - **Behavior**:
>     - Supports loading rules from global (`~/.roo/rules/`) and project-local (`.roo/rules/`) directories.
>     - Project-specific rules override global rules.
>     - Supports mode-specific rules via `rules-{mode}/` directories.
>     - Falls back to legacy `.roorules` files if no `.roo/rules` directories exist.
>   - **Architecture**:
>     - Adds `roo-config` service with functions: `getGlobalRooDirectory()`, `getProjectRooDirectoryForCwd()`, `directoryExists()`, `fileExists()`, `readFileIfExists()`, `getRooDirectoriesForCwd()`, `loadConfiguration()`.
>     - Updates `custom-instructions.ts` to use `roo-config` service for rule loading.
>   - **Tests**:
>     - Adds tests for `loadRuleFiles()` and `addCustomInstructions()` in `custom-instructions-global.spec.ts`.
>     - Adds tests for `roo-config` functions in `index.spec.ts`.
> 
> <sup>This description was created by </sup>[<img alt="Ellipsis" src="https://img.shields.io/badge/Ellipsis-blue?color=175173">](https://www.ellipsis.dev?ref=RooCodeInc%2FRoo-Code&utm_source=github&utm_medium=referral)<sup> for fa190a90fb44ea1d999f639f35e14697463486a3. You can [customize](https://app.ellipsis.dev/RooCodeInc/settings/summaries) this summary. It will automatically update as commits are pushed.</sup>

<!-- ELLIPSIS_HIDDEN -->

daniel-lxs commented on src/services/roo-config/index.ts:
Should we consider moving the global rules to the extension's global storage directory? This would align with how global MCPs and custom modes are handled and could improve consistency.

daniel-lxs commented on src/core/prompts/sections/custom-instructions.ts:
```diff
@@ -155,19 +157,33 @@ function formatDirectoryContent(dirPath: string, files: Array<{ filename: string
+		if (await directoryExists(rulesDir)) {
+			const files = await readTextFilesFromDirectory(rulesDir)
+			if (files.length > 0) {
+				const isGlobal = rooDir.includes(path.join(os.homedir(), ".roo"))
```
The current implementation uses `rooDir.includes(path.join(os.homedir(), ".roo"))` to detect if a directory is global. This could be fragile if the home directory path contains ".roo" elsewhere. Have you considered using an exact path comparison instead?

daniel-lxs commented on src/services/roo-config/index.ts:
```diff
@@ -0,0 +1,122 @@
+export async function readFileIfExists(filePath: string): Promise<string | null> {
+	try {
+		return await fs.readFile(filePath, "utf-8")
+	} catch {
```
The `loadConfiguration` function currently catches all errors when reading files. Would it be better to only catch expected errors, like `ENOENT`, to avoid masking other potential issues?

daniel-lxs commented on src/core/prompts/sections/__tests__/custom-instructions-global.spec.ts:
```diff
@@ -0,0 +1,230 @@
+
+			const result = await loadRuleFiles(mockCwd)
+
+			originalConsoleLog("Result:", result)
```
The tests include `originalConsoleLog` calls that don't seem to be used in the implementation. Could these be removed or converted to test assertions for cleaner tests?

daniel-lxs commented on src/services/roo-config/index.ts:
```diff
@@ -0,0 +1,122 @@
+}
+
+// Export with backward compatibility alias
+export const loadRooConfiguration = loadConfiguration
```
The `loadRooConfiguration` alias at the end of `roo-config/index.ts` appears to have an incomplete type annotation.

daniel-lxs commented on src/services/roo-config/index.ts:
```diff
@@ -0,0 +1,122 @@
+}
+
+/**
+ * Loads configuration from multiple .roo directories with project overriding global
```
Would it be helpful to add examples of the expected directory structure in the JSDoc comments to guide users?

daniel-lxs reviewed (commented):
Thanks for your contribution! This PR is a great addition, providing a much-needed hierarchical configuration system. I've left a few suggestions for improvement, mostly around consistency and robustness. Overall, great work!

daniel-lxs reviewed (approved):
Thank you for addressing the suggestions!

I made a small change to clean up the rules headers to follow follow the current behavior, I don't think there's a benefit to explaining where the rules come from 2 times:

![image](https://github.com/user-attachments/assets/39408dbd-1616-436b-aaa5-113705907968)

samhvw8 commented:
> Thank you for addressing the suggestions!
> 
> I made a small change to clean up the rules headers to follow follow the current behavior, I don't think there's a benefit to explaining where the rules come from 2 times:
> 
> ![image](https://private-user-images.githubusercontent.com/57051444/458428777-39408dbd-1616-436b-aaa5-113705907968.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTA3NzUyMDIsIm5iZiI6MTc1MDc3NDkwMiwicGF0aCI6Ii81NzA1MTQ0NC80NTg0Mjg3NzctMzk0MDhkYmQtMTYxNi00MzZiLWFhYTUtMTEzNzA1OTA3OTY4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA2MjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNjI0VDE0MjE0MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWM5MWRkM2QwYTI2ZjIyNDg4NDk3MDMwMzE5M2ZiYmIwYWY4NWM2ZWZkYmQ2YzJhZDE3ODI4NzI3MDAxODYzYzMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.fNUQAU0vXw5-0URpUoQv6ClwvEj9UVmop21yvSTRZCc)

thanks you for polish it <3
